lib_eio/mbox.ml

```diff
@@ -0,0 +1,50 @@
+type 'a t = {
+  writers: unit Waiters.t;
+  readers: unit Waiters.t;
+  mutex: Mutex.t;
+  id: Ctf.id;
+  value: 'a option ref;
+}
+
+let create () = {
+  writers = Waiters.create ();
+  readers = Waiters.create ();
+  mutex = Mutex.create ();
+  id = Ctf.mint_id ();
+  value = ref None;
+}
+
+let send (t:'a t) (m : 'a) : unit =
+  let sent = ref false in
+  let wait () =
+    Mutex.lock t.mutex;
+    match !(t.value) with
+    | Some _ ->
+      Waiters.await ~mutex:(Some t.mutex) t.writers t.id
+    | None ->
+      t.value := Some m;
+      Waiters.wake_one' t.readers ();
+      sent := true;
+      Mutex.unlock t.mutex
```

Comment:
wouldn't you want to unlock before waking up the reader? (same below for writer)

Reply:
Not really, could be before the `sent := true` obviously, readers are in a wait-queue protected by the same mutex as writers.
`Waiters.await` passes the lock-down and that protects the internal `Lwt_dllist`, both wait-queues are under the same lock.

Reply:
I pushed a diff to do the `sent := true` and `received := true` outside the lock, but the `wakeup_one` call needs to be locked.

Reply:
Ok, makes sense. I suppose that `wakeup_one` will not _immediately_ wake the fiber, it will be queued? otherwise the worry was that it'd try to lock the mutex just before it's released…

Reply:
> Ok, makes sense. I suppose that `wakeup_one` will not _immediately_ wake the fiber, it will be queued? otherwise the worry was that it'd try to lock the mutex just before it's released…

Correct, the `wakeup_one` resolves the `Suspended` effect created in `Waiters.await`, enqueues it into the `run-queue`, so next time the current `Fiber` blocks, it will pop the next one from the `run-queue` (at least that's as far as my understanding of the code goes at this time, I haven't really dived into the scheduler).

Reply:
Doing some more digging, I believe `wakeup_one` should be different, so that we don't end up calling the `enqueue` callback while holding the mutex, something like:
```ocaml
Mutex.lock t.mutex;
let waker = Waiters.pop_next_waiter t in
Mutex.unlock t.mutex;
Waiters.wake waker
```

Reply:
There is also the risk of a deadlock, possibly, if the reader this just poped from `waiters` turns out to have been cancelled while it slept? I think? I'm new to all this. With threads I've often started with `Condition.signal` only to move to `Condition.broadcast`.

Reply:
> There is also the risk of a deadlock, possibly, if the reader this just poped from `waiters` turns out to have been cancelled while it slept? I think? I'm new to all this. With threads I've often started with `Condition.signal` only to move to `Condition.broadcast`.

Leme make sure I understood:
```ocaml
Mutex.lock t.mutex;
let waker = Waiters.pop_next_waiter t in
Mutex.unlock t.mutex;
(* at this point you mean the waker has been canceled *)
Waiters.wake waker
```

So I end up not waking the next one in queue, which in turn deadlocks ?

Reply:
I pushed a change, my `wakeup_one'` is broken, I tried to describe the possible deadlock in the commit:
https://github.com/ocaml-multicore/eio/pull/334/commits/161579fbf982ead4e1b601d80cecc617b9dbf281

Thanks a lot for this ! I really appreciate you making me ponder about it !

Reply:
:+1: